backend-rs/src/utils.rs

```diff
@@ -0,0 +1,425 @@
+        peer_connection.set_local_description(answer).await?;
+        ice_gather_rx.recv().await;
+
+        Ok(peer_connection
+            .local_description()
+            .await
+            .ok_or("Failed to Generate Description")?)
+    }
+}
+
+/*
+    How CameraThreadHandle Works:
+
+    This struct manages a "reader" thread that owns an H264CameraReader. Importantly, this struct sends WebRTC client Receivers
+    to the reader thread. The reader thread then constantly loops and reads H264 bytes from H264CameraReader and then iterates
+    through all the WebRTC client Receivers, sending each receiver a copy of the bytes read. If there aren't any more Receivers,
+    then the reader thread is dropped and CameraThreadHandle is removed from WebcamManager.
+
+    This could be thought of as an "event system" in which all the WebRTC clients on the tokio runtime receiver bytes from the reader thread.
+*/
+
+pub struct CameraThreadHandle {
+    camera_path: String,
+    cam_mode_tx: Sender<CameraMode>,
+    current_mode: CameraMode,
+    camera_modes: Vec<CameraMode>,
+    manual_shutdown_needed: Arc<AtomicBool>,
+    sink_flush_needed: Arc<AtomicBool>,
+    tx_sink: Arc<Mutex<Vec<Sender<Vec<u8>>>>>,
+}
+
+impl CameraThreadHandle {
+    fn start_camera_thread(
+        camera_path: &str,
+        camera_handles: Arc<Mutex<Vec<CameraThreadHandle>>>,
+    ) -> Result<Self, Box<dyn Error>> {
+        let camera_path = camera_path.to_owned();
+        let (cam_mode_tx, mut cam_mode_rx) = mpsc::channel::<CameraMode>(1);
+        let manual_shutdown_needed: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));
+        let sink_flush_needed: Arc<AtomicBool> = Arc::new(AtomicBool::new(false));
+        let tx_sink: Arc<Mutex<Vec<Sender<Vec<u8>>>>> = Arc::new(Mutex::new(Vec::new()));
+
+        let c_camera_path = camera_path.clone();
+        let c_manual_shutdown_needed = manual_shutdown_needed.clone();
+        let c_sink_flush_needed = sink_flush_needed.clone();
+        let c_tx_sink = tx_sink.clone();
+
+        let device_path = Path::new(&c_camera_path);
+        let node = v4l::context::enum_devices()
+            .into_iter()
+            .find(|node| node.path() == device_path)
+            .ok_or("V4l Node Not Found")?;
+        let mut device = Device::new(node.index())?;
+        let modes = CameraMode::fetch_all(&device)?;
+        let initial_mode = *modes.last().ok_or("No Last")?; // The last camera mode tends to be the one with the best resolution and fps.
+        thread::spawn(move || {
+            let mut reader = H264CameraReader::new(&mut device, initial_mode).unwrap();
+            let mut rtc_txs: Vec<Sender<Vec<u8>>> = Vec::new();
+
+            loop {
+                // Shutdown due to drop of CameraThreadHandle
+                if c_manual_shutdown_needed.load(Ordering::SeqCst) {
+                    return;
+                }
+
+                // Used to prevent having to constantly lock the c_tx_sink mutex.
```

Comment:
This might be 2% more idiomatic? ;D


```suggestion
            while !c_manual_shutdown_needed.load(Ordering::SeqCst) {
```

Reply:
lol not sure why it replaced the comment. it swaps out the `loop {`